lib/rubygems/commands/push_command.rb

```diff
@@ -98,24 +89,11 @@ def send_gem(name)
       terminate_interaction 1
     end
 
-    gem_data = Gem::Package.new(name)
-
-    unless @host
-      @host = gem_data.spec.metadata['default_gem_server']
-    end
-
-    push_host = nil
-
-    if gem_data.spec.metadata.has_key?('allowed_push_host')
-      push_host = gem_data.spec.metadata['allowed_push_host']
-    end
-
-    @host ||= push_host
+    _, push_host = get_hosts_for(name)
 
-    # Always include @host, even if it's nil
     args += [ @host, push_host ]
 
-    say "Pushing gem to #{@host || Gem.host}..."
+    say "Pushing gem to #{@host}..."
```

Comment:
I think `@host` could actually be `nil` if... everything is `nil`. If that's the case, it _does_ default to `Gem.host` unless `Gem.configuration.disable_default_gem_server` is true.

I think I would keep this message and the previous comment (`# Always include @host, even if it's nil`) as they are now, and we can iterate the refactoring further later.